centreon/src/Core/ResourceAccess/Application/UseCase/AddRule/AddRule.php

```diff
@@ -0,0 +1,429 @@
+            ['ruleId' => $ruleId, 'contact_group_ids' => $rule->getLinkedContactGroupIds()]
+        );
+
+        $this->writeRepository->linkContactGroupsToRule($ruleId, $rule->getLinkedContactGroupIds());
+    }
+
+    /**
+     * This method will ensure that the ids provided for the dataset filters are valid
+     * and that the structure of the dataset is correct regarding the hierarchy defined in the DatasetFilter entity.
+     *
+     * @param AddRuleRequest $request
+     *
+     * @throws RuleException
+     * @throws \InvalidArgumentException
+     *
+     * @return DatasetFilter[]
+     */
+    private function validateAndCreateDatasetFiltersFromRequest(AddRuleRequest $request): array
+    {
+        $datasetFilters = [];
+
+        $validateAndBuildDatasetFilter = function (
+            array $data,
+            ?DatasetFilter $parentDatasetFilter
+        ) use (&$validateAndBuildDatasetFilter, &$datasetFilter): void {
+            // In any case we want to make sure that resources provided are valid
+            // Then at DatasetFilter creation the hierarchy is validated
+            $this->validator->assertIdsAreValid($data['type'], $data['resources']);
+
+            // first iteration we want to create the root filter
+            if ($datasetFilter === null) {
+                $datasetFilter = new DatasetFilter(
+                    type: $data['type'],
+                    resourceIds: $data['resources']
+                );
+                if ($data['dataset_filter'] !== null) {
+                    $validateAndBuildDatasetFilter($data['dataset_filter'], null);
+                }
+            } else {
+                // we want to create the first children
+                if ($parentDatasetFilter === null) {
+                    $filter = new DatasetFilter(type: $data['type'], resourceIds: $data['resources']);
+                    $datasetFilter->setDatasetFilter($filter);
+                    if ($data['dataset_filter'] !== null) {
+                        $validateAndBuildDatasetFilter($data['dataset_filter'], $datasetFilter->getDatasetFilter());
+                    }
+                } else {
+                    // create any children available and assign it to the parent
+                    $childrenDatasetFilter = new DatasetFilter(
+                        type: $data['type'],
+                        resourceIds: $data['resources']
+                    );
+
+                    $parentDatasetFilter->setDatasetFilter($childrenDatasetFilter);
+
+                    if ($data['dataset_filter'] !== null) {
+                        $validateAndBuildDatasetFilter($data['dataset_filter'], $childrenDatasetFilter);
+                    }
+                }
+            }
+        };
+
+        foreach ($request->datasetFilters as $dataset)  {
+            $datasetFilter = null;
+            $validateAndBuildDatasetFilter($dataset, $datasetFilter);
+
+            /** @var DatasetFilter $datasetFilter */
```

Comment:
Forcing th type by removind the null possibility can lead to design problems.
It would be better to design an algorithm with a return of a `DatasetFilter`.
We should also avoid the reference inside the closure.

Reply:
That will be done later on